### ISSUE TITLE
Clarity and/or discussion on "private" property/method naming

### ISSUE THREAD

**Issue by ChrisDahmCS** [angular/angular#27760]
# 📚 Docs or angular.io bug report

### Description

I was hoping to gain more clarity or talk about a "nice to have" in regards to the [Angular Styleguide suggestion](https://angular.io/guide/styleguide#properties-and-methods) not to use the underscore prefix for "private" properties and methods if possible. Specifically, it's use and the way "private" interacts from ts within templates between the default dev and prod builds - it seems like there could be a gap here.

[This question ](https://github.com/angular/angular/issues/27615) helps illustrate part of the issue, and has been talked about at length in a few more references (and more), [here_1](https://github.com/angular/angular/issues/26393), [here_2](https://github.com/angular/angular/issues/16893).

While these talk more specifically about access methods conflicting with prop names, I wanted to bring up the interaction with templates. Issues [like this one](https://github.com/angular/angular/issues/27615), while they are not a problem with the framework, throwing an error that a property is "private" while looking at the template currently requires extra step(s) ([explained more here](https://github.com/angular/angular/issues/11978#issuecomment-255736505)).

While JavaScript (default ng serve / ES5) does not care about "private," tsc (default prod build) does.TS has support for letting the developer know if a private property or method is being used, but I'm not familiar with a way of seeing this in the templates. Developers who are more familiar with Angular may not have an issue here, but using a pattern like this could be a "nice to have" for newer/unfamiliar developers, or a timesaving naming/visual pattern when working in templates.

### What's the affected URL?**
https://angular.io/guide/styleguide#properties-and-methods

### Reproduction Steps**
```
ng new test-ng
cd test-ng
```
Inside of app.component.ts:
```
  public test = 'test';
  private testPrivate = 'testPrivate';
  private _testPrivate = '_testPrivate';
```
and inside of app.component.html:
```
<div>
  {{test}}
</div>
<div>
  {{testPrivate}}
</div>
<div>
  {{_testPrivate}}
</div>
```

Running the default `ng serve` works just fine as expected / noted above, but `ng build --prod` will fail (also as expected). The difference (at least for me) is that with the IDEs I've used, I can immediately tell by looking just at the template that `_testPrivate` probably is private- 

### Expected vs Actual Behavior**

Actual behavior: while it's an agreed upon mistake by the developer, no ahead of time warning for private props/methods in templates, requires additional steps.

Expected: Can be a range from noting the alternative naming with underscore, adding more detail to the existing info of TS vs JS "private", or leave the docs alone and provide/suggest tooling out of the box that catches this at dev time.

**Comment by trotyl** [comment]
> as we re-implementing the template type-check in ivy

@IgorMinar TypeScript access modifiers are lexical, all static properties as well as the initialization codes are considered part of the class and has no problem to access those properties, so direct component template would already works fine.

The problem now is `<ng-template>`s will be extracted to standalone functions that outside the lexical scope of class body, then losing the accessibility to private class. (Not happened in the initial Ivy implementation, but latter changes)

To make it work, we can use an IIFE to wrap all `<ng-template>` and keep it side class body like:

```typescript
class AppComponent {
  private foo: string

  static ngComponentDef = defineComponent({
    template: (() => {
      function div_Template_1(rf: RenderFlags, ctx: XXX) {
        // Not sure how `nextContext` would be typed yet, possibly by generics?
        nextContext(0).foo;
      }

      return function AppComponent_Template(rf: RenderFlags, ctx: AppComponent) {
        // reference to div_Template_1
      }
    })()
  })
}
```

**Comment by IgorMinar** [comment]
@trotyl I actually proposed the same thing yesterday while writing up the comment above, but then we discussed it with @mhevery and realized that this won't work because the element to component lookup is happening at runtime, so the type-checker cannot check if binding to a property on an element is valid or not, because at this time we don't know if the element is a component that has given property declared as an input.

All of this can be solved by not relying on css selectors to identify components in the template, but that is a much bigger project, one that we are deferring until Ivy ships.

**Comment by trotyl** [comment]
@IgorMinar I think you‘re discussing type-checking for Inputs rather than accessing private property, which is a different story, the context consumption is static, no matter the target being Input, property or invalid thing.

**Comment by pkozlowski-opensource** [comment]
> All of this can be solved by not relying on css selectors to identify components in the template, but that is a much bigger project, one that we are deferring until Ivy ships.

Exactly my thinking. As such I would propose updating style guide (1) and do "the right thing" post-ivy. As you've noted: 

> is that the component is forced to exposes more api surface than it should, but for anyone who uses AOT this is already something they do, so it would not be a regression, but rather admitting the current state of things

this is what everyone is doing already since this is what the rendering engine implementation expects.

**Comment by mlc-mlapis** [comment]
@trotyl ... with relation to accessing private properties ... how **Public class fields** (Chrome 72) and **Private class fields** (very close also) ... https://developers.google.com/web/updates/2018/12/class-fields will affect it? Probably it's possible to expect that TS will compile private modifiers  to private class fields ... `#count = 0;`.

**Comment by mlc-mlapis** [comment]
@trotyl ... thanks.

**Comment by trotyl** [comment]
@IgorMinar @pkozlowski-opensource I'm not sure why Angular team decide on not allowing `private` access, but that's only a design choice, not technical limitation. In current Ivy implementation it can be simply move the code generation position:

https://github.com/angular/angular/blob/30d1f292c930d0ba8c198757b3333d4b594692d8/packages/compiler-cli/src/ngtsc/typecheck/src/context.ts#L265-L266

from:

```typescript
class SomeCmp {}

function _tcb_1(ctx: SomeCmp) {
  // generated here
}
```

to:

```typescript
class SomeCmp {
  static _tcb_1(ctx: SomeCmp) {
    // generated here
  }
}
```

The only requirement here is making `private`/`protected` accessor part of [`requiresInlineTypeCheckBlock`](https://github.com/angular/angular/blob/30d1f292c930d0ba8c198757b3333d4b594692d8/packages/compiler-cli/src/ngtsc/typecheck/src/type_check_block.ts#L829) criteria.

Given there's no real cost for supporting `private` access, can you share the reason why not support this in post-Ivy infrastructure?

**Comment by MaKCbIMKo** [comment]
Is there any updates on that?

I'm really interested in using access modifiers other than `public` within component/directive in order to not expose its internal API.

**Comment by pkozlowski-opensource** [comment]
I agree that private members of component and their visibility in a template should be clarified. Marking as confusing aspects of the framework so we can update docs and / or think of a better mental model.
